Disabling or deleting Expired status breaks membership status update

agh1 opened the pull request:
Overview
----------------------------------------
The `Expired` membership status is not reserved.  That means people can disable, rename, or delete it.  However, if it was deleted, disabled, or renamed, the status update would break.

Before
----------------------------------------
If you disabled the `Expired` membership status, the "Membership status processor" scheduled job would fail with the following message:

> Finished execution of Membership status processor with result: Failure, Error message: A fatal error was triggered: One of parameters (value: ) is not of the type Integer


After
----------------------------------------
Disabling, deleting, or renaming the status doesn't cause problems.

Technical Details
----------------------------------------
In lieu of rearranging the query when there is no `Expired` status, I simply substitute `0` because it's an integer and there should be no status with that ID.

Comments
----------------------------------------
The presumption that expired memberships should not be evaluated for their status is questionable.  This doesn't address that.

eileenmcnaughton commented:
@agh1 & you don't think it SHOULD be reserved?

agh1 commented:
@eileenmcnaughton I don't think it should be reserved.  The only bit of code relying on it existing is this function here that just doesn't bother checking if expired members deserve a different status.

On the flip side, the ship has sailed on making it reserved.  Who knows how many organizations have edited or deleted the status?  Re-introducing it isn't trivial; it would have to fit in with each organization's configuration.  Finally, some organizations (including three of our clients, all coming to this independently) don't have a concept of automatic expiration: members are either canceled or delinquent (but still treated as members until they pro-actively cancel).

mattwire commented:
@davejenx This is similar to stuff you've worked on before. Are you able to review?

eileenmcnaughton commented on CRM/Member/BAO/Membership.php:
```diff
@@ -2241,7 +2241,7 @@ public static function updateAllMembershipStatus() {
 
     $updateCount = $processCount = self::updateDeceasedMembersStatuses();
-    $allStatus = CRM_Member_PseudoConstant::membershipStatus();
+    $allStatus = CRM_Member_BAO_Membership::buildOptions('status_id', 'get');
```
So looking at https://docs.civicrm.org/dev/en/latest/framework/pseudoconstant/

![Screenshot 2019-03-18 14 51 36](https://user-images.githubusercontent.com/336308/54502054-99bc9480-498d-11e9-8db0-d9a8587ae191.png)

I see a risk - if someone relabels expired this will not pick it up - I think we should use 'validate' rather than 'get' as the context & in the unit test we should set the name & label to not be the same to test that.

(Note that there are multiple other places on the BAO where the label is checked rather than the name - I think fixing those is out of scope but perhaps it would be good to add todos to them)

agh1 commented:
@eileenmcnaughton I agree--the pseudoconstant function returns names, and I accidentally got labels.  I changed it to `validate` and fixed the tests to find the statuses by name.  The other spots in this BAO that use the buildOptions with `get` do so in order to display the label in plain text, so they're fine.

While I was at it I noticed a reference to CRM-15475 which is sort of similar, just with disabled statuses. 
 It was resolved in #4492 by doing a separate psuedoconstant call, but since this change takes care of both renamed and disabled statuses, I made `Cancelled` work the same as the other statuses.

agh1 commented:
Test failure is `api_v3_LoggingTest::testGet`:

```
Exception: CRM_Core_Exception: "No match found for key : to with value : Dwarf, DopeyArray
```

I have no idea why this would fail.  It's passing locally.

eileenmcnaughton commented:
@agh1 agree test fail is unrelated. I'm just wondering if we can make one more tweak in the unit test & set the label to something other than 'Expired' so we are protected against some later change breaking the label bit

eileenmcnaughton commented on tests/phpunit/CRM/Member/BAO/MembershipStatusTest.php:
```diff
@@ -131,6 +131,14 @@ public function testDel() {
   public function testExpiredDisabled() {
+    $result = civicrm_api3('MembershipStatus', 'get', [
+      'name' => "Expired",
+      'api.MembershipStatus.create' => ['label' => 'Expiiiired'],
```
lol - thanks!

agh1 replied:
:ghost: don't mark me `Deceased` I'm just `Expiiiired` :ghost:

eileenmcnaughton replied:
:-)
